sklearn/linear_model/tests/test_coordinate_descent.py

```diff
@@ -111,7 +113,10 @@ def test_lasso_zero():
     # Check that the lasso can handle zero data without crashing
     X = [[0], [0], [0]]
     y = [0, 0, 0]
-    clf = Lasso(alpha=0.1).fit(X, y)
+    # _cd_fast.pyx tests for gap < tol, but here we get 0.0 < 0.0
+    # should probably be changed to gap <= tol ?
```

Comment:
Indeed, please feel free to open a dedicated follow-up PR once this one is merged.

Reply:
Actually I think this should be fixed as part of this PR to simplify it.

Reply:
Also once this is done, could you please update this PR to add inline comments to explain why we expect a given warning each time we ignore it in the tests?

Reply:
Thanks for the comments, I will push a change for `_cd_fast.pyx` but want to do some testing with that (first non-python only change).

Concerning the comments: that is a good point. My idea is that what we are checking here is that two different methods to do a fitting (pipeline or intern scaling for example) result in the same fitting, not whether the fitting actually converges.

But I will  **try** to check for each case to find an explanation ...

Reply:
@ogrisel Ok, that turned out to be a bit more tricky and takes time (at least for me): If I change the code in `_cd_fast.pyx` to do
```
@@ -819,7 +819,7 @@ def enet_coordinate_descent_multi_task(
                 gap += l1_reg * l21_norm - const * ry_sum + \
                      0.5 * l2_reg * (1 + const ** 2) * (w_norm ** 2)
 
-                if gap < tol:
+                if gap <= tol:
                     # return if we reached desired tolerance
                     break
         else:
```
then suddenly a warning that *should* (?) be there disappeared: the test `test_enet_coordinate_descent` in `test_coordinate_descent.py` has the following comment:
```
    """Test that a warning is issued if model does not converge"""
```
and is tested against `Lasso` (`precompute = True | False`) and against `MultiTaskLasso` (interestingly, two times the same test), and now suddenly `MultiTaskLasso` converges while `Lasso` does not.

The code that is run is basically
```
X = np.ones((5, 2)) * 1e50
y = np.ones((5,2))
MultiTaskLasso(max_iter=2).fit(X,y)
```
Since this boils down to looking at what is going on in `_cd_fast.pyx`, function `enet_coordinate_descent_multi_task` I need a bit of time to understand how that works.

If any of you have an insight at hand, please let me know.

Reply:
BTW, I dropped the duplication of `MultiTaskLasso` in the last commit pushed, but that is cosmetics.

Reply:
Can you please push the change, either as part of this PR or in a new draft PR to see what is failing with specific error message and so that we can have a more focused discussion on particular failures caused by this change?

Reply:
Hi @ogrisel I pushed the `gap <= tol` change to the PR (and a typo fix).